tortoisebeacon/handlers.go

```diff
@@ -25,92 +26,99 @@ const TBFollowingVotingProtocol = "TBFollowingVotingGossip"
 var (
 	// ErrMalformedProposal is returned if proposal message is malformed.
 	ErrMalformedProposal = errors.New("malformed proposal message")
+
+	// ErrProposalDoesntPassThreshold is returned if proposal message doesn't pass threshold.
+	ErrProposalDoesntPassThreshold = errors.New("proposal doesn't pass threshold")
 )
 
 // HandleSerializedProposalMessage defines method to handle Tortoise Beacon proposal Messages from gossip.
 func (tb *TortoiseBeacon) HandleSerializedProposalMessage(ctx context.Context, data service.GossipMessage, sync service.Fetcher) {
+	receivedTime := time.Now()
+
 	tb.Log.With().Debug("New proposal message",
 		log.String("from", data.Sender().String()))
 
-	var m ProposalMessage
-	if err := types.BytesToInterface(data.Bytes(), &m); err != nil {
+	var message ProposalMessage
+	if err := types.BytesToInterface(data.Bytes(), &message); err != nil {
 		tb.Log.With().Error("Received malformed proposal message",
 			log.String("message", string(data.Bytes())),
 			log.Err(err))
 
 		return
 	}
 
-	if err := tb.handleProposalMessage(ctx, m); err != nil {
-		tb.Log.With().Error("Failed to handle proposal message",
-			log.String("sender", data.Sender().String()),
-			log.String("message", m.String()),
-			log.Err(err))
+	currentEpoch := tb.currentEpoch()
+	if message.EpochID < currentEpoch {
+		tb.Log.With().Debug("Received proposal message from previous epoch, ignoring",
+			log.Uint64("message_epoch", uint64(message.EpochID)),
+			log.Uint64("current_epoch", uint64(currentEpoch)))
 
 		return
 	}
 
-	data.ReportValidation(ctx, TBProposalProtocol)
-}
+	tb.proposalChansMu.Lock()
+	ch := tb.getOrCreateProposalChannel(message.EpochID)
+	tb.proposalChansMu.Unlock()
 
-func (tb *TortoiseBeacon) handleProposalMessage(ctx context.Context, m ProposalMessage) error {
-	// TODO(nkryuchkov): buffer messages that came from future epochs because of time drifts on different nodes (https://github.com/spacemeshos/go-spacemesh/pull/2529#issuecomment-877772213)
+	extendedMessage := extendedProposalMessage{
+		message:      message,
+		gossip:       data,
+		receivedTime: receivedTime,
+	}
 
-	receivedTimestamp := time.Now()
+	select {
+	case <-tb.CloseChannel():
+	case ch <- extendedMessage:
+	}
+}
+
+func (tb *TortoiseBeacon) handleProposalMessage(m ProposalMessage, receivedTime time.Time) error {
 	currentEpoch := tb.currentEpoch()
 
 	atxID, err := tb.atxDB.GetNodeAtxIDForEpoch(m.MinerID, currentEpoch-1)
-	if err != nil {
+	if errors.Is(err, database.ErrNotFound) {
```

Comment:
I think you need to add the minerID and current epoch to debug log -this will help finding issues if there is a consensus failure